src/Microsoft.OData.Core/DefaultServicesConfigurations.cs

```diff
@@ -0,0 +1,68 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="DefaultServicesConfigurations.cs" company="Microsoft">
+// Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.OData.Edm;
+using Microsoft.OData.Json;
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.OData.Core
+{
+    /// <summary>
+    /// Extension methods for <see cref="IServiceCollection"/>.
+    /// </summary>
+    public static class DefaultServicesConfigurations
+    {
+        /// <summary>
+        /// Adds the default OData services to the <see cref="IServiceCollection"/>.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/> to add the services to.</param>
+        /// <returns>The <see cref="IServiceCollection"/> instance itself.</returns>
+        public static IServiceCollection AddDefaultODataServices(this IServiceCollection services)
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            return services.AddDefaultODataServices(ODataVersion.V4);
+        }
+
+        /// <summary>
+        /// Adds the default OData services to the <see cref="IServiceCollection"/>.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/> to add the services to.</param>
+        /// <param name="odataVersion">ODataVersion for the default services.</param>
+        /// <returns>The <see cref="IServiceCollection"/> instance itself.</returns>
+        public static IServiceCollection AddDefaultODataServices(this IServiceCollection services, ODataVersion odataVersion)
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            services.AddSingleton<IJsonReaderFactory, DefaultJsonReaderFactory>();
+            services.AddSingleton<IJsonWriterFactory, DefaultJsonWriterFactory>();
+            services.AddSingleton<IJsonWriterFactoryAsync, DefaultJsonWriterFactory>();
+            services.AddSingleton(sp => ODataMediaTypeResolver.GetMediaTypeResolver(null));
+            services.AddScoped<ODataMessageInfo>();
+            services.AddSingleton(sp => new ServicePrototype<ODataMessageReaderSettings>(new ODataMessageReaderSettings()));
+            services.AddScoped(sp => sp.GetServicePrototype<ODataMessageReaderSettings>().Clone());
+            services.AddSingleton(sp => new ServicePrototype<ODataMessageWriterSettings>(new ODataMessageWriterSettings()));
+            services.AddScoped(sp => sp.GetServicePrototype<ODataMessageWriterSettings>().Clone());

```

Comment:
I think we should consider dropping the use of `ServicePrototype` and just register new instances as scoped directly. But I guess that should be a separate PR discussion.

@xuzhg @gathogojr looked at this code and we couldn't understand why it was designed this way. And it may be potentially related to some race conditions that occur due to some shared settings state being modified concurrently by separate requests.

Reply:
@habbes  I saw prototypes are used when instantiation of a service requires a lot of resources or is resource intensive. A prototype is created which is then cloned in subsequent calls so that the whole instantiation of the service is not done over and over again. I agree more discussion is required here.